### PR TITLE
Autolog with PyTorch using lightning - Examples

### PR DESCRIPTION
**Pull request by shrinath-suresh**
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

## What changes are proposed in this pull request?

This PR is dependent on https://github.com/mlflow/mlflow/pull/3601 . Adding MNIST handwritten recognition and also Bert news classification as examples for pytorch lightning autolog

## How is this patch tested?

Tested by running the examples and checking the results in mlflow server.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: examples/pytorch/BertNewsClassification/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+## Bert Sentiment Analysis example with MLFlow
```
Comment (shrinath-suresh) [inline examples/pytorch/BertNewsClassification/README.md:1]: Bert News Classification. I will change the description here

File: examples/pytorch/BertNewsClassification/MLproject [hunk 2]
```diff
@@ -0,0 +1,22 @@
+name: bert-sentiment
```
Comment (shrinath-suresh) [inline examples/pytorch/BertNewsClassification/MLproject:1]: will change sentiment to classification

File: examples/pytorch/BertNewsClassification/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+This is an experimental api that logs ML model artifacts and metrics.
+The metrics are logged during training of the model.
+
+## Setup
+
+Mlflow tracking UI is set to `http://localhost:5000` with experiment name as `Default` in the script - `bert_classification.py`
+Change the variables according to the environment setup.
+
+### Running the code
+To run the example via MLflow, navigate to the `mlflow/examples/pytorch/SentimentAnalysis` directory and run the command
```
Comment (shrinath-suresh) [inline examples/pytorch/BertNewsClassification/README.md:20]: ditto

File: examples/pytorch/MNIST/example2/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
+## MNIST example with MLFlow
```
Comment (smurching) [inline examples/pytorch/MNIST/example2/README.md:1]: @shrinath-suresh what's the difference between this example (example2) and example1? Could we include just one or the other?

Comment (smurching) [reply examples/pytorch/MNIST/example2/README.md:1]: Also, is this example adapted from an existing PytorchLightning example somewhere? Maybe we could reference the existing example code, and mention ("This example is adapted from ... with only minor tweaks, i.e. the addition of an `mlflow.pytorch.autolog` call to add automatic instrumentation of training code")

File: examples/pytorch/MNIST/example2/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+In this example, we train a model to predict handwritten digits.The autolog code uses Pytorch Lightning's MLFlowLogger to log metrics.
```
Comment (smurching) [inline examples/pytorch/MNIST/example2/README.md:3]:
```suggestion
In this example, we train a Pytorch Lightning model to predict handwritten digits. The code is almost entirely dedicated to model training, with the addition of a single ``mlflow.pytorch.autolog()`` call to enable automatic logging of params, metrics, and models from training. 
```

I think we could reword the overview as in the above suggestion^

File: examples/pytorch/MNIST/example2/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
+## MNIST example with MLFlow
+
+In this example, we train a model to predict handwritten digits.The autolog code uses Pytorch Lightning's MLFlowLogger to log metrics. 
+The code is trained using pytorch lightning loop and the autolog function call in the main - `autolog()`
+is responsible for logging the params, metrics, model summary and the model.
+This example logs metrics only after n epoch iterations. The iteration limit can be set in the autolog method using the parameter `log_every_n_iter=NUMBER-OF-ITERATIONS`.
+For ex: `autolog(log_every_n_iter=5)`
```
Comment (smurching) [inline examples/pytorch/MNIST/example2/README.md:7]: Could we remove this in favor of linking to the ``mlflow.pytorch.autolog`` doc once it's available, as described in https://github.com/mlflow/mlflow/pull/3608/files#r514601992?

File: examples/pytorch/MNIST/example2/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+### Code related to MLflow:
+* [`mlflow.pytorch.pytorch_autolog`]
+This is an experimental api that logs ML model artifacts and metrics.
+The metrics are logged during training of the model.
```
Comment (smurching) [inline examples/pytorch/MNIST/example2/README.md:12]:
```suggestion
```

I think we can remove this section given the rewording suggested in https://github.com/mlflow/mlflow/pull/3608/files#r514601992

File: examples/pytorch/MNIST/example2/README.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+## Setting tracking URI
```
Comment (smurching) [inline examples/pytorch/MNIST/example2/README.md:14]:
```suggestion
## Logging to a custom tracking server

To configure MLflow to log to a custom (non-default) tracking location, set the ``MLFLOW_TRACKING_URI`` environment variable, e.g. via  ``export MLFLOW_TRACKING_URI=http://localhost:5000/``.  For more details, see [the docs](https://mlflow.org/docs/latest/tracking.html#where-runs-are-recorded)
```

Could we reword this section as suggested? And move it to the end of the README? I think it'll only be relevant to a small number of readers

File: examples/pytorch/MNIST/example2/README.md (hunk 4 above)
Comment (smurching) [inline examples/pytorch/MNIST/example2/README.md:1]: cc @andreakress would you be able to help take a look at the README here for wording?

**Review by smurching — COMMENTED** [review]
@shrinath-suresh could we add tests for the new examples under https://github.com/mlflow/mlflow/blob/master/tests/examples/test_examples.py? We should be able to add to the existing tests, which allow for running a Python script or MLflow project and asserting that it's successful.

In particular, you can add a test for running the example as a Python script by adding a new line with test args [here](https://github.com/mlflow/mlflow/blob/64c93c81167b8a132bf00535fc23f3420c9caf7f/tests/examples/test_examples.py#L113), and a test for running the example as an MLflow project by adding new test args [here](https://github.com/mlflow/mlflow/blob/64c93c81167b8a132bf00535fc23f3420c9caf7f/tests/examples/test_examples.py#L58)

File: examples/pytorch/MNIST/example2/README.md (hunk 4 above)
Comment (shrinath-suresh) [reply examples/pytorch/MNIST/example2/README.md:1]: > @shrinath-suresh what's the difference between this example (example2) and example1? Could we include just one or the other?

Example 1 illustrates - EarlyStopping Behaviour, Model Checkpointing, Learning Rate Monitor
Example 2 illustrates - Logging every n iterations

Comment (shrinath-suresh) [reply examples/pytorch/MNIST/example2/README.md:1]: > Also, is this example adapted from an existing Pytorch Lightning example somewhere? Maybe we could reference the existing example code, and mention ("This example is adapted from ... with only minor tweaks, i.e. the addition of an `mlflow.pytorch.autolog` call to add automatic instrumentation of training code")

sure. i will add a link to it

**Comment by chauhang** [comment]
> @shrinath-suresh could we add tests for the new examples 

@smurching We will do a separate PR for the Examples tests as per our discussion offline

File: examples/pytorch/MNIST/README.md [hunk 9]
```diff
@@ -0,0 +1,3 @@
+## MNIST Handwritten Digit Recognition examples
+
+This folder contains two subdirectories: `example1`, which demonstrates MNIST autolog example with Early stopping and model checkpoint, `example2`, demonstrates MNIST autolog example where metrics are logged every n iterations.
```
Comment (smurching) [inline examples/pytorch/MNIST/README.md:3]: I'm ok with merging this as is in this PR, but in a follow-up could we just combine these two examples into a single one? It seems we could show logging every N iterations via a single commented-out line of code in ``example1`` (e.g. have a commented-out line like):

```
# Replace the existing mlflow.pytorch.autolog() call with the one below to log training metrics
# every 10 epochs, instead of the default of logging on every epoch
# mlflow.pytorch.autolog(every_n_epoch=10)
mlflow.pytorch.autolog()
```

**Review by smurching — APPROVED** [review]
LGTM to get the actual example code in for now, given that we're planning to add example tests and add clarifying READMEs in follow-up PRs

**Comment by tchaton** [comment]
Hey @shrinath-suresh,

I had plenty of comments for this PR 👍 It got merged sightly too fast ahaha

Screenshot 2020-11-04 at 08 39 45
Screenshot 2020-11-04 at 08 39 54
Screenshot 2020-11-04 at 08 40 01
Screenshot 2020-11-04 at 08 40 06
Screenshot 2020-11-04 at 08 40 14
Screenshot 2020-11-04 at 08 40 20
Screenshot 2020-11-04 at 08 40 31
Screenshot 2020-11-04 at 08 40 38

Best regards,
Thomas Chaton.

**Comment by tchaton** [comment]
Hey @smurching,

Please, ping me to review the next PR :)

Best regards,
Thomas Chaton.
Research Engineer at Pytorch Lightning.

**Comment by shrinath-suresh** [comment]
@tchaton Thanks for your review comments. This PR is a base version of examples. We already are in process of creating a follow up PR for this. We will address the review comments and add you as reviewer.

**Comment by tchaton** [comment]
Hey @shrinath-suresh,

Great ! I will add more reviewers from Lightning side to make sure your examples are using the latest API :)

Best regards,
Thomas Chaton.

**Comment by chauhang** [comment]
@tchaton We will review your feedback and see what elements will be applicable in the Autolog context. Please sync-up with William Falcon at your end, he is aware of this project. 

Autologing is adding support for automatically logging the metrics during training runs. You can check the [doc pages build](https://2069-136202695-gh.circle-artifacts.com/0/docs/build/html/tracking.html#automatic-logging) from the document PR. 

Best regards,
Geeta
